allocation.py

```diff
@@ -22,13 +27,31 @@ def get_rec_name(self, name):
     def search_rec_name(cls, name, clause):
         return [('employee.rec_name',) + tuple(clause[1:])]
 
+    @fields.depends('work', '_parent_work.company')
+    def on_change_with_company(self, name=None):
+        if self.work and self.work.company:
+            return self.work.company.id
+
+    @classmethod
+    def search_company(cls, name, clause):
+        return [('work.company',) + tuple(clause[1:])]
+
 
 class Work(metaclass=PoolMeta):
     __name__ = 'project.work'
     allocations = fields.One2Many('project.allocation', 'work', 'Allocations')
     employees = fields.Function(fields.Char('Employees'), 'get_employees',
         searcher='search_employees')
 
+    @classmethod
+    def __setup__(cls):
+        super(Work, cls).__setup__()
+        readonly = Eval('allocations', [0])
+        if 'readonly' in cls.company.states:
```

Comment:
Not allow change company when hava allocation lines because each allocation is related to the company.